In [21]:
import numpy as np
import tensorflow as tf
from tensorflow import keras
import matplotlib.pyplot as plt
from tensorflow.keras.losses import SparseCategoricalCrossentropy
from tensorflow.keras.datasets import cifar10
from keras import layers
from tensorflow.keras.preprocessing.image import ImageDataGenerator

In [22]:
# Load data from cifar10 dataset
(X_train_orig, Y_train_orig), (X_test_orig, Y_test_orig) = cifar10.load_data()

In [23]:
#Normalize image matrix
X_train = X_train_orig/255.0
X_test = X_test_orig/255.0

Y_train = Y_train_orig
Y_test = Y_test_orig

print ("number of training examples = " + str(X_train.shape[0]))
print ("number of test examples = " + str(X_test.shape[0]))
print ("X_train shape: " + str(X_train.shape))
print ("Y_train shape: " + str(Y_train.shape))
print ("X_test shape: " + str(X_test.shape))
print ("Y_test shape: " + str(Y_test.shape))

number of training examples = 50000
number of test examples = 10000
X_train shape: (50000, 32, 32, 3)
Y_train shape: (50000, 1)
X_test shape: (10000, 32, 32, 3)
Y_test shape: (10000, 1)


In [24]:
model1 = tf.keras.Sequential([
                tf.keras.layers.Conv2D(filters=64, kernel_size=(3,3), activation='relu', input_shape=(32,32, 3), strides=(1,1), name='conv0'),
                tf.keras.layers.MaxPooling2D((2, 2), name='max_pool0'),
        
                tf.keras.layers.Conv2D(filters=128, kernel_size=(3,3), activation='relu', strides=(1,1), name='conv1'),
                tf.keras.layers.MaxPooling2D(pool_size=(2, 2), name='max_pool1'),
        
                tf.keras.layers.Flatten(),
                tf.keras.layers.Dense(128, activation='relu', name='fc1'),
                tf.keras.layers.Dense(10, activation='softmax', name='fc2'),
            ])

In [25]:
model1.compile(loss=SparseCategoricalCrossentropy(), metrics=['accuracy'])

In [26]:
model1.fit(X_train, Y_train, epochs=5)

Epoch 1/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 26s 16ms/step - accuracy: 0.4008 - loss: 1.6614
Epoch 2/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 26s 17ms/step - accuracy: 0.6458 - loss: 1.0193
Epoch 3/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 26s 16ms/step - accuracy: 0.7129 - loss: 0.8306
Epoch 4/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 26s 16ms/step - accuracy: 0.7600 - loss: 0.6984
Epoch 5/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 25s 16ms/step - accuracy: 0.7910 - loss: 0.6103


In [27]:
model1.evaluate(X_test, Y_test)

313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 8ms/step - accuracy: 0.7147 - loss: 0.9123


[0.9207584857940674, 0.7127000093460083]

In [28]:
data_augmentation = keras.Sequential([
    layers.RandomFlip("horizontal"),           # Flip images horizontally
    layers.RandomRotation(0.1),                # Randomly rotate images by ±10%
    layers.RandomZoom(0.1)                     # Randomly zoom images
])

In [29]:
model2 = tf.keras.Sequential([
                data_augmentation,
                tf.keras.layers.Conv2D(filters=64, kernel_size=(3,3), activation='relu', input_shape=(32,32, 3), strides=(1,1), name='conv0'),
                tf.keras.layers.MaxPooling2D((2, 2), name='max_pool0'),
        
                tf.keras.layers.Conv2D(filters=128, kernel_size=(3,3), activation='relu', strides=(1,1), name='conv1'),
                tf.keras.layers.MaxPooling2D(pool_size=(2, 2), name='max_pool1'),
        
                tf.keras.layers.Flatten(),
                tf.keras.layers.Dense(128, activation='relu', name='fc1'),
                tf.keras.layers.Dense(10, activation='softmax', name='fc2'),
            ])

In [30]:
model2.compile(loss=SparseCategoricalCrossentropy(), metrics=['accuracy'])

In [31]:
model2.fit(X_train, Y_train, epochs=5)

Epoch 1/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 32s 20ms/step - accuracy: 0.3533 - loss: 1.7715
Epoch 2/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 32s 20ms/step - accuracy: 0.5523 - loss: 1.2705
Epoch 3/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 32s 21ms/step - accuracy: 0.5961 - loss: 1.1544
Epoch 4/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 32s 20ms/step - accuracy: 0.6205 - loss: 1.0961
Epoch 5/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 33s 21ms/step - accuracy: 0.6391 - loss: 1.0455


In [32]:
model2.evaluate(X_test, Y_test)

313/313 ━━━━━━━━━━━━━━━━━━━━ 3s 9ms/step - accuracy: 0.6211 - loss: 1.1101


[1.1273961067199707, 0.6154000163078308]

### Conclusion

As observed from the results:

- The **first model** (without data augmentation) achieved a training accuracy of **0.7910**, but due to overfitting, its test accuracy dropped sharply to **0.7147**. This significant drop indicates weaker generalization.

- The **second model** (with data augmentation) achieved a training accuracy of **0.6391** and a test accuracy of **0.6211**. The smaller drop suggests that this model is **less prone to overfitting** and generalizes more effectively.

**Summary Comparison:**  
- Model 1: Training = 0.7910 → Testing = 0.7147 (large drop)  
- Model 2: Training = 0.6391 → Testing = 0.6211 (smaller drop)

**Final Remark:**  
Although Model 2 has a lower overall accuracy, it is better suited for **generalization**, making it more robust for real-world scenarios.
